### PR TITLE
REFACTOR: Unite Your Requirements!

### PR DESCRIPTION
**Pull request by akmalsoliev**
With the arrival of a lightning-fast package manager, mighty UV, the requirements.txt is like yesterday's news. Instead of navigating through a jungle of dependencies, like a mismatched sock drawer, everything can now unite in one glorious spot—just like Italy in 1861!

This marvelous change even makes installing dependencies blazingly fast (except for cudf—polars is better anyways). Now it's as easy as pie with just one command: uv sync --all-extras.  No more typing three commands when you can do it in one! 1 > 3 — gives you more time to enjoy that delicious Italian caffè.

## What type of PR is this? (check all applicable)

- [x] 💾 Refactor
- [ ] ✨ Feature
- [ ] 🐛 Bug Fix
- [ ] 🔧 Optimization
- [ ] 📝 Documentation
- [ ] ✅ Test
- [ ] 🐳 Other

## Checklist

- [ ] Code follows style guide (ruff)
- [ ] Tests added 
- [ ] Documented the changes

### CODE REVIEW

**Comment by akmalsoliev** [comment]
P.S. I've updated `pyproject.toml` and banished those pesky `.txt` files. More changes are coming; just testing the waters before diving in!

File: pyproject.toml [hunk 1]
```diff
@@ -31,13 +31,40 @@ exclude = [
...
+    "mkdocstrings[python]>=0.26.1",
+    "polars>=1.0.0",
+]
+#cudf = ["cudf>=23.08.00"]
 modin = ["modin"]
 pandas = ["pandas>=0.25.3"]
 polars = ["polars>=0.20.3"]
 pyarrow = ["pyarrow>=11.0.0"]
-dask = ["dask[dataframe]>=2024.7"]
+dask = ["dask[dataframe]>=2023.5.0"]
```
Comment (FBruzzesi) [inline pyproject.toml:67]: Maybe Marco recalls why that was the min version

**Review by FBruzzesi — COMMENTED** [review]
Hey there! Thanks for your PR. We definitly started to be opinionated to suggest uv in our contribution guide. However I have mixed feelings about this one. Let me try to mention a few reason (sorted by dev experience):

- If someone's workflow is not on uv, then how can he/she/they pick up dev-dependencies?
- `uv sync` is _too_ strict for my (but I guess many) development settings and to avoid that  it doesn't uninstall the rest of the libraries in the env, I need to remember to run `uv sync --inexact` instead
- For a few set up based on docker (e.g. devcontainers and codespaces), having a file allows to easily copy and install the dependencies _before_ even copying the python package. This leads to easier caching (at least in my experience).

**Comment by FBruzzesi** [comment]
> Embracing change can indeed be challenging, and that's perfectly normal.

Definitely, and we would like to let people use whatever workflow works for them (not only pip) as well. Thanks for adjusting the pyproject section.

> From what I understand, Docker caching should remain intact whether using `pyproject.toml` or `requirements.txt`. If the current setup is causing issues, I'm keen to explore alternatives that work better for you. Here’s an example to illustrate:

> ```
> FROM python:latest
> 
> COPY ./pyproject.toml /pyproject.toml
> 
> RUN pip install .
> 
> RUN echo installing dev && pip install .[dev]
> 
> RUN echo installing polars && pip install .[polars]
> ```
> 
> Please let me know if there's anything specific you find challenging, and I’d be happy to discuss further.

The above snippet would run from `RUN pip install .` each time something changes in the codebase.
What I usually do is something along the lines:

```dockerfile
...
COPY ./requirements.txt /requirements.txt

RUN uv pip install -r requirements.txt # this will be cached until something changes in requirements.txt

COPY <rest of codebase> <rest of codebase>

RUN uv pip install .
...
```

There should be a way to link requirements.txt dynamically into pyproject, so personally that's how I would do it but happy to discuss

**Comment by MarcoGorelli** [comment]
thanks - you'll need to update the instructions for this to work (e.g. ci, contributing)

fancy making a PR first, so that github automatically runs ci for you, and then coming back to this?

**Comment by akmalsoliev** [comment]
Ciao @FBruzzesi, 
Sorry for late reply, what you think about this setup? 

```docker
FROM ghcr.io/astral-sh/uv:latest

COPY ./pyproject.toml pyproject.toml

RUN uv venv

RUN uv pip compile pyproject.toml --extra dev > requirements.txt
RUN uv pip install -r requirements.txt

RUN echo hello 

RUN uv pip install -r pyproject.toml --extra docs

COPY . .

RUN uv sync --all-extras 
```

When trying on the local machine it felt like there was a caching of 
```
RUN uv pip compile pyproject.toml --extra dev > requirements.txt
RUN uv pip install -r requirements.txt
```

File: pyproject.toml (hunk 1 above)
Comment (FBruzzesi) [reply pyproject.toml:65]: Not sure why this was resolved, but 2024.7 is indeed the min version we can support, as we need dask_expr.

File: pyproject.toml [hunk 2]
```diff
@@ -32,12 +32,37 @@ exclude = [
...
+    "pyarrow>=17.0.0",
+    "pytest>=8.3.3",
+    "pytest-cov>=5.0.0",
+    "pytest-env>=1.1.5",
+    "pytest-randomly>=3.15.0",
+    "scikit-learn>=1.3.2",
+    "tqdm>=4.66.5",
+    "typing-extensions>=4.12.2",
+    "dask[dataframe]>=2023.5.0"
+]
```
Comment (FBruzzesi) [inline pyproject.toml:51]: For running in ci/cd we should not pin any of these, as we want to have some old and random versions as well. Also dask requires python 3.9 or above.

File: pyproject.toml [hunk 3]
```diff
@@ -32,12 +32,37 @@ exclude = [
...
+docs = [
+    "jinja2>=3.1.4",
+    "markdown-exec[ansi]>=1.9.3",
+    "mkdocs>=1.6.1",
+    "mkdocs-material>=9.5.36",
+    "mkdocstrings[python]>=0.26.1",
+    "polars>=1.0.0",
+]
+#cudf = ["cudf>=23.08.00"]
```
Comment (FBruzzesi) [inline pyproject.toml:60]: Any reason for commenting this one out? 🤔

**Review by FBruzzesi — COMMENTED** [review]
Hey @akmalsoliev thanks again for the effort. The workflow you are proposing seems to work quite well and being able to cache layers properly in docker.

Given that is sorted out, I left some comments that probably require attention and, as @MarcoGorelli mentioned, we need this to be incorporated in the ci/cd workflows properly.
If possible including nox and the contributing documentation as well!

Let me know if I can help in any way forward 😉

File: pyproject.toml (hunk 3 above)
Comment (akmalsoliev) [reply pyproject.toml:60]: Hi, sorry for the delayed response. If I recall correctly, @MarcoGorelli brought up this issue regarding installing cuDF with conda. The installation process is quite complex, and I'm not sure how to resolve it.
image

File: pyproject.toml (hunk 2 above)
Comment (akmalsoliev) [reply pyproject.toml:51]: not an issue, was more of demonstration, will be fixed

**Pull request closed without merging**